### ISSUE TITLE
CheckClassAdapter severe slowdown for TestExamples

### ISSUE THREAD

**Issue by magnus-madsen** [flix/flix#188]
Using the CheckClassAdapter causes a significant slowdown for the TestExamples (as observed by profiling).

Questions:
- Is this to be expected? (i.e. working as intended, but checking is slow)
- Other than disabling the checker, can something be done?
- Should we hide the checker behind `-Xdebug` or similar? 

Actionable items:
- Add codegen (and other phases) to `Time`. 

I think if we fix this, then #178 may also be fixed (Although the underlying problem is unknown).

**Comment by mhyee** [comment]
Is it `CheckClassAdapter` or `ClassWriter`? How much of a slowdown?

---

The documentation doesn't really say whether `CheckClassAdapter` slows things down, but I don't think it's surprising. I don't think we can do anything other than disabling it.

That might not be such a bad idea, if we assume the tests are good enough to catch issues. `CheckClassAdapter` doesn't really help the user (since they'll get an error anyway, if invalid code is generated).

In fact, `CheckClassAdapter` makes it harder to debug invalid code, because it throws an error and doesn't output any bytecode. I have to disable the adapter if I want to look at the invalid bytecode.

---

`ClassWriter` is a different story, especially since we enabled it to compute stack frames (which also computes stack/variable maximums). The documentation explicitly states that computing stack heights causes a 10% slowdown, while computing stack frames causes a **100%** slowdown. ([See page 44](http://download.forge.objectweb.org/asm/asm4-guide.pdf).)

Apparently, the library uses a general algorithm which is very slow. There may be a faster algorithm for our specific situation (which we would have to implement manually), but it may also be the case that our situation requires the general and slow algorithm.

There's one other thing we could try. From informal observations, the numbers on [this line](https://github.com/flix/flix/blob/6373afe27e6a44acf74133c96c493d525be5b63d/main/src/ca/uwaterloo/flix/language/phase/Codegen.scala#L243) seem to make a difference (or maybe only if there's an error). These are fake values for the max stack/variables, set to the JVM maximums. Reducing these numbers would probably speed things up (since the library is restricted to smaller stacks), but it does put a limit the Flix programs that can be written.

---

To summarize, there are two things you could try right now:

1. Disable `CheckClassAdapter`
2. Reduce the values in [this line](https://github.com/flix/flix/blob/6373afe27e6a44acf74133c96c493d525be5b63d/main/src/ca/uwaterloo/flix/language/phase/Codegen.scala#L243).

Beyond these options, we would have to look into calculating the stack frames ourselves.

**Comment by magnus-madsen** [comment]
The CheckClassAdapter takes 9.6s out of 16s total.

**Comment by magnus-madsen** [comment]
Proposal:

- Ensure that every phase of the compiler is represented in `Time`. Once this is done, we can get a better understanding of where the compiler is spending its time.
- Add a flag `-Xinvariants` (or alternatively `-Xassertions`) which toggles compiler checks, e.g. in this case, the flag would enable the use of the `CheckClassAdapter`. (Notice how this flag is separate from `-Xdebug` which should mean to generate/output as much debugging information as possible, e.g. by dumping ASTs, bytecode, etc.)

**Comment by magnus-madsen** [comment]
![image](https://cloud.githubusercontent.com/assets/2287953/16777594/2639d6f2-4838-11e6-833e-e61f39776ee2.png)

**Comment by magnus-madsen** [comment]
Added the `--Xinvariants` flag. Without it, you get:

![image](https://cloud.githubusercontent.com/assets/2287953/16780059/f7e0abf4-4842-11e6-87e6-36bcc5bc3831.png)

:-)

**Comment by mhyee** [comment]
I'm actually surprised. I would have thought that calculating the stack frames would have been much more expensive.

Does this resolve #178?

**Comment by magnus-madsen** [comment]
Fixed. But it is unclear if we still have build issues. Continuing in #178 .
